### ISSUE TITLE
New RedundantBraces is too aggressive

### ISSUE THREAD

**Issue by olafurpg** [scalameta/scalafmt#1147]
This template is a guideline, not a strict requirement.
- **Version**: 1.5.0
- **Integration**: cli
- **Configuration**:
```
rewrite.rules = [
  RedundantBraces
]
```

## Steps

Given code like this:
```scala

      case FormatToken(open @ LeftParen(), _, _) if {
            val isSuperfluous = isSuperfluousParenthesis(open, leftOwner)
            leftOwner match {
              case _: Term.If | _: Term.While | _: Term.For | _: Term.ForYield
                  if !isSuperfluous =>
                true
              case _ => false
            }
          } =>
```

When I run scalafmt like this:

```bash
scalafmt
```

## Problem

Scalafmt formats code like this:
```scala
>
...
[warn] Error in /Users/ollie/dev/scalafmt/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala: <input>:1036: error: illegal start of simple expression
val isSuperfluous = isSuperfluousParenthesis(open, leftOwner)
            ^
```

## Expectation
I expect the code to format since it has no parse errors.

## Workaround

```
rewrite.redundantBraces.generalExpressions = false
```

## Notes

Regression since https://github.com/scalameta/scalafmt/pull/1122

**Comment by olafurpg** [comment]
I don't know if this is intended behavior or not, but I was surprised to see

```diff
-    while (curr.left.start < end.start && curr != prev) {
-      if (matches(curr.left)) {
-        result += curr.left
-        goToMatching()
-      } else {
-        prev = curr
-        curr = next(curr)
-      }
+    while (curr.left.start < end.start && curr != prev) if (matches(curr.left)) {
+      result += curr.left
+      goToMatching()
+    } else {
+      prev = curr
+      curr = next(curr)
     }
```

Question if curly braces should be kept around while loops and if statements without an else 🤔

**Comment by olafurpg** [comment]
Infix operators are a tough one as well, 


```diff
-            (opensConfigStyle(formatToken) || {
-              forceConfigStyle(leftOwner) && !styleMap.forcedBinPack(leftOwner)
-            }) =>
+            (opensConfigStyle(formatToken) ||
+              forceConfigStyle(leftOwner) && !styleMap.forcedBinPack(leftOwner)) =>
```

It's technically redundant but operator precedence and associativity is difficult sometimes so guiding braces may help readability

**Comment by olafurpg** [comment]
Opened https://github.com/scalameta/scalafmt/pull/1157 to disable `generalExpressions` by default until this is fixed to unblock v1.5.1

**Comment by olafurpg** [comment]
I took a stab at fixing this in https://github.com/olafurpg/scalafmt/compare/general-expression?expand=1 but there are some complications to distinguish between case bodies and case guards.


@japgolly would you be interested in taking a stab at this?
